flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,314 @@
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FilesCommitter implements Committer<FilesCommittable>, Serializable {
+  private static final long serialVersionUID = 1L;
+  private static final long INITIAL_CHECKPOINT_ID = -1L;
+  private static final Logger LOG = LoggerFactory.getLogger(FilesCommitter.class);
+  private static final String FLINK_JOB_ID = "flink.job-id";
+  private static final int MAX_RECOMMIT_TIMES = 3;
+
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed-checkpoint-id";
+
+  // TableLoader to load iceberg table lazily.
+  private final TableLoader tableLoader;
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final Table table;
+  private transient Map<String, Long> maxCommittedCheckpointIdForJob = Maps.newHashMap();
+
+  private transient ExecutorService workerPool;
+
+  public FilesCommitter(
+      TableLoader tableLoader,
+      boolean replacePartitions,
+      Map<String, String> snapshotProperties,
+      int workerPoolSize) {
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool(
+            "iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<FilesCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    Collection<CommitRequest<FilesCommittable>> store = Lists.newArrayList();
+    String jobId = null;
+    long maxCommittableCheckpointId = INITIAL_CHECKPOINT_ID;
+
+    for (CommitRequest<FilesCommittable> request : requests) {
+      WriteResult committable = request.getCommittable().committable();
+      Long committableCheckpointId = request.getCommittable().checkpointId();
+      jobId = request.getCommittable().jobID();
+      if (request.getNumberOfRetries() > MAX_RECOMMIT_TIMES) {
+        String message =
+            String.format(
+                "Failed to commit transaction %s after retrying %d times",
+                request.getCommittable(), MAX_RECOMMIT_TIMES);
+        request.signalFailedWithUnknownReason(new RuntimeException(message));
+      } else if (request.getNumberOfRetries() > 0) {
+        commitResult(
+            committable.dataFiles().length,
+            committable.deleteFiles().length,
+            Lists.newArrayList(request));
+      } else if (committableCheckpointId
+              > maxCommittedCheckpointIdForJob.getOrDefault(jobId, Long.MAX_VALUE)
+          || getMaxCommittedCheckpointId(table, jobId) == -1) {
```

Comment:
Maybe we want to calculate the `getMaxCommittedCheckpointId(table, jobId)` once per loop.
Also we have to make sure that the transaction is only committed if the Table has not been changed since between the time we get the `maxCommitterCheckpointId` and the commit. We have to make sure that commit is not automatically retried by some mechanism if there is a conflict

Reply:
Maybe I didn't quite catch what you mean.
But one thing, I think, is that the table can be changed by other job between the time we get the `maxCommitterCheckpointId` and the commit. For a flink job, the JobId is unique, parallel jobs do not use the same jobid for writing. 

When multiple Flink tasks are written in parallel, Iceberg itself ensures the transactional for commit actions. When it commits a conflict, it tries again.

For `committableCheckpointId> maxCommittedCheckpointIdForJob.getOrDefault(jobId, Long.MAX_VALUE)
 || getMaxCommittedCheckpointId(table, jobId) == -1`:

In theory, `maxCommittedCheckpointIdForJob` it will contain only at most two key, one is the job id of the current job, one is the `pre-job-id` of the pre-job.
No other task will have the same pre-job-id at the same time. So, we think that if `maxCommittedCheckpointIdForJob`  does not contain the jobid of this request , and also no corresponding records in the table, then, that means
The jobid has never written data to the table. At this point, whether we are restore from the pre-job or it is the first time of this job,  will accumulate the data and commit them all at once later.

Reply:
Is it possible to commit some files in the pre-job, and commit some files in the actual job?

My previous understanding was, that we commit only in the committer, and we just collect all of the written files, calculate the min and the max commitId, check that it is not in conflict with the current Iceberg table data, and if everything is ok, then we commit them in one Iceberg commit.

Reading your code I had to change my mental model a bit, because I understand now, that commits with deletes should be committed separately, so we need to order the commits by commitId, and collect the commits until we find a commit with delete files. In this case, we have to commit the previous batch, and then commit to current one, and then we can continue going through the rest.

The other part of my review comment was referring to the ways how Iceberg handles commits atomicity and retries. Iceberg detects conflicts. When we start an Iceberg transaction we lock on an Iceberg snapshot (the actual snapshot of the table object we use to create the change), and if the table is changed before the Iceberg commit call, we will see a different current snapshotId for the table, and have an error.
So that is why we have to be careful when we refresh the table and when we retrieve the maxCommitId.
Also there are ways for Iceberg where the code tries to automatically recover conflicts. If the files does not conflict, then it tries to commit the changes with new metadata without notifying the caller (at least I have read about this somewhere). This retry logic might not be desirable for us. We have to check if this conflict handling will cause issues with our maxCommitId handing, or not

Reply:
> Reading your code I had to change my mental model a bit, because I understand now, that commits with deletes should be committed separately, so we need to order the commits by commitId, and collect the commits until we find a commit with delete files. In this case, we have to commit the previous batch, and then commit to current one, and then we can continue going through the rest.

👍  Because of the particularity of the v2 row-level delete, it is processed separately.

> The other part of my review comment was referring to the ways how Iceberg handles commits atomicity and retries.

If Iceberg fails to commit the request, we do not retry it. So, I've streamlined the code here, and when a commit fails, we `signalFailedWithKnownReason` or `signalFailedWithUnknownReason`.  This way, when Iceberg fails to commit. If iceberg itself tries again and still fails, it makes no sense for us to retry the commit again and may produce errors.

In the recovery case, when the job is recovered from the pre, it immediately attempts to commit the data of the pre-job. Therefore, the data of the current job does not coexist with the pre.